bundles/org.openhab.binding.smartthings/src/main/java/org/openhab/binding/smartthings/internal/SmartthingsHandlerFactory.java

```diff
@@ -55,9 +55,10 @@
  * @author Bob Raker - Initial contribution
  */
 @NonNullByDefault
-@Component(service = { ThingHandlerFactory.class,
+@Component(service = { ThingHandlerFactory.class, SmartthingsHubCommand.class,
```

Comment:
If you're fixing this. Why not fix it properly and move all code that should be in the bridge to the brigde. So it becomes much simpler as you don't need to keep track of the things here as they're already known by the bridge and also don't have to pass bridge information around here.

Reply:
Sorry, I don’t understand what would be proper here. Let me explain why it is the way it is and maybe you can explain what is wrong.

SmartthingsHandlerFactory;sendDeviceCommand() is dependent on the httpClient service. To my knowledge I have to get access to this in a component. Therefore sendDeviceCommand has to be in a component or at least be somewhere with the httpClient service can be passed to it. I could move that to the bridge but I don’t see how that makes things simpler.

SmartthingsHandlerFactory;sendDeviceCommand() is used from SmartthingsDiscoveryService and SmartthingsThingHandler. In SmartthingsDiscoveryService how do I get a reference to the Bridge?

Thanks

Reply:
@Hilbrand 
I'm still waiting for feedback and more direction on what you think the proper solution is.

Reply:
Because discovery depends on a bridge. It means it should be handled differently than just making it a component. This can be implemented by let the discovery implements `ThingHandlerService` See this prefix documentation on how that can be implemented: https://deploy-preview-1262--openhab-docs-preview.netlify.app/docs/developer/bindings/#discovery-that-is-bound-to-a-thing That implementation gives you access to the bridge in the discovery class and then you can call methods from the bridge. This also means the methods in the factory that do something with the bridge can be moved to the bridge class.

Reply:
@Hilbrand 
I've implemented the changes in the document that you provided but now I am getting compile errors that I don't know how to fix.  Can you point me to a working binding that uses this approach?

Here is what I've done and the resulting errors:
**Discovery Service**
`@NonNullByDefault
@Component(service = { DiscoveryService.class,
        EventHandler.class }, configurationPid = "discovery.smartthings", property = "event.topics=org/openhab/binding/smartthings/discovery")
public class SmartthingsDiscoveryService extends AbstractDiscoveryService implements DiscoveryService, ThingHandlerService, EventHandler {

... existing code ...
    @Override
    public void setThingHandler(@Nullable ThingHandler handler) {
        if (handler instanceof SmartthingsBridgeHandler) {
            bridgeHandler = (SmartthingsBridgeHandler) handler;
        }
    }

    @Override
    public @Nullable ThingHandler getThingHandler() {
        return bridgeHandler;
    }
}`
Compiler messages:
The inherited method AbstractDiscoveryService.deactivate() cannot hide the public abstract method in ThingHandlerService
Redundant superinterface DiscoveryService for the type SmartthingsDiscoveryService, already defined by AbstractDiscoveryService

**Bridge Handler**
`
  @NonNullByDefault
  public class SmartthingsBridgeHandler extends ConfigStatusBridgeHandler implements ThingHandlerService {

... existing code ...

    @Override
    public Collection<Class<? extends ThingHandlerService>> getServices() {
        return Collections.singleton(SmartthingsDiscoveryService.class);
    }`
Compiler messages
The type SmartthingsBridgeHandler must implement the inherited abstract method ThingHandlerService.getThingHandler()
The type SmartthingsBridgeHandler must implement the inherited abstract method ThingHandlerService.setThingHandler(ThingHandler)

Please advise on how I can move forward. At this point I am answering questions from users about when the discovery service will work.  I would like to move this forward. Soon.

Reply:
In DiscoveryService you should override the deactivate method:
```java
    @Override
    public void deactivate() {
        super.deactivate();
    }
```

The class `SmartthingsBridgeHandler` should not implements `ThingHandlerService`.

Ps. I've edited your comment to use code blocks.